In [1]:
import pandas as pd
import numpy as np
import datetime as dt

# Turn off warnings to ease reading
import warnings
warnings.filterwarnings('ignore')

# Loading datasets

In [2]:
vgc_data = pd.read_csv('Scraping/vgsales.csv', index_col = 0)

In [3]:
rawg_data = pd.read_csv('Scraping/rawg_games_smaller.csv', parse_dates = ['released','updated'])

In [4]:
vgc_data.head()

Title Platform       Publisher  \
0            Tetris   Series         Unknown   
1       Super Mario   Series        Nintendo   
2           Pokemon   Series        Nintendo   
3  Grand Theft Auto   Series  Rockstar Games   
4              FIFA   Series       EA Sports   

                               Developer  VGC_Score  Critic_Score  User_Score  \
0                        Alexey Pajitnov        NaN           NaN         NaN   
1                               Nintendo        NaN           NaN         NaN   
2                             Game Freak        NaN           NaN         NaN   
3                         Rockstar North        NaN           NaN         NaN   
4  Extended Play Productions (1991-1997)        NaN           NaN         NaN   

  Total_Shipped Total_Sales NA_Sales EU_Sales JP_Sales Other_Sales  \
0       496.00m         NaN      NaN      NaN      NaN         NaN   
1       372.86m         NaN      NaN      NaN      NaN         NaN   
2       369.88m         NaN      NaN      NaN      NaN         NaN   
3       335.00m         NaN      NaN      NaN      NaN         NaN   
4       325.00m         NaN      NaN      NaN      NaN         NaN   

       Release  Last_Update  
0  06th Jun 84  27th Feb 20  
1  20th Jul 83  20th Feb 20  
2  28th Sep 98  03rd Feb 20  
3  27th Mar 98  03rd Feb 20  
4  15th Dec 93  03rd Feb 20

In [5]:
rawg_data.head()

id                        slug                         name   released  \
0  3498          grand-theft-auto-v           Grand Theft Auto V 2013-09-17   
1  4200                    portal-2                     Portal 2 2011-04-18   
2  3328     the-witcher-3-wild-hunt     The Witcher 3: Wild Hunt 2015-05-18   
3  5286                 tomb-raider           Tomb Raider (2013) 2013-03-05   
4  5679  the-elder-scrolls-v-skyrim  The Elder Scrolls V: Skyrim 2011-11-11   

     tba  metacritic  suggestions_count    updated  \
0  False        97.0                416 2021-03-03   
1  False        95.0                582 2020-08-03   
2  False        92.0                678 2020-10-02   
3  False        86.0                664 2020-08-03   
4  False        94.0                621 2020-07-06   

                                           platforms  \
0  ['PC', 'Xbox Series S/X', 'PlayStation 5', 'Pl...   
1  ['Xbox One', 'PlayStation 3', 'PC', 'Xbox 360'...   
2  ['PC', 'Xbox One', 'Nintendo Switch', 'PlaySta...   
3  ['PC', 'PlayStation 4', 'PlayStation 3', 'Xbox...   
4  ['PC', 'PlayStation 3', 'Xbox 360', 'Nintendo ...   

                           genres  \
0         ['Action', 'Adventure']   
1           ['Shooter', 'Puzzle']   
2  ['Action', 'Adventure', 'RPG']   
3         ['Action', 'Adventure']   
4               ['Action', 'RPG']   

                                              stores  \
0  ['Epic Games', 'PlayStation Store', 'Xbox Stor...   
1  ['Xbox Store', 'Xbox 360 Store', 'PlayStation ...   
2  ['GOG', 'Xbox Store', 'Steam', 'PlayStation St...   
3  ['App Store', 'Google Play', 'PlayStation Stor...   
4  ['Xbox 360 Store', 'Nintendo Store', 'Steam', ...   

                                                tags   esrb_rating  
0  ['Singleplayer', 'Steam Achievements', 'Multip...        Mature  
1  ['Singleplayer', 'Steam Achievements', 'Multip...  Everyone 10+  
2  ['Singleplayer', 'Atmospheric', 'Full controll...        Mature  
3  ['Singleplayer', 'Multiplayer', 'Atmospheric',...        Mature  
4  ['Singleplayer', 'Steam Achievements', 'steam-...        Mature

# Cleaning vgc_data

There are columns we will not use:

- Sales  columns except `Total_Shipped` and `Total_Sales`.

- `Last_Update`.

- `VGC_Score`, `Critic_Score` and `User Score`, as we are going to get scores from Metacritic.



In [6]:
vgc_data = vgc_data.drop(['VGC_Score','Critic_Score','User_Score','NA_Sales','EU_Sales','JP_Sales','Other_Sales','Last_Update'], axis = 1)
vgc_data.head()

Title Platform       Publisher  \
0            Tetris   Series         Unknown   
1       Super Mario   Series        Nintendo   
2           Pokemon   Series        Nintendo   
3  Grand Theft Auto   Series  Rockstar Games   
4              FIFA   Series       EA Sports   

                               Developer Total_Shipped Total_Sales  \
0                        Alexey Pajitnov       496.00m         NaN   
1                               Nintendo       372.86m         NaN   
2                             Game Freak       369.88m         NaN   
3                         Rockstar North       335.00m         NaN   
4  Extended Play Productions (1991-1997)       325.00m         NaN   

       Release  
0  06th Jun 84  
1  20th Jul 83  
2  28th Sep 98  
3  27th Mar 98  
4  15th Dec 93

We are going to transform `Total_Shipped` and `Total_Sales` to numeric.

In [7]:
vgc_data['Total_Shipped'] = vgc_data['Total_Shipped'].str.strip('m').astype('float')
vgc_data['Total_Sales'] = vgc_data['Total_Sales'].str.strip('m').astype('float')

---
Now we proceed to clean the `Release` column.

We can observe the following:

- There are NaNs
- Days are in the ~st ~nd ~rd ~th format
- Months are in 3 letter format
- Years have the last 2 digits format

In [8]:
sorted(vgc_data['Release'].str[-2:].dropna().unique(), reverse = True)

['99',
 '98',
 '97',
 '96',
 '95',
 '94',
 '93',
 '92',
 '91',
 '90',
 '89',
 '88',
 '87',
 '86',
 '85',
 '84',
 '83',
 '82',
 '81',
 '80',
 '79',
 '78',
 '77',
 '75',
 '73',
 '71',
 '70',
 '21',
 '20',
 '19',
 '18',
 '17',
 '16',
 '15',
 '14',
 '13',
 '12',
 '11',
 '10',
 '09',
 '08',
 '07',
 '06',
 '05',
 '04',
 '03',
 '02',
 '01',
 '00']

We can see that:
- The years belonging to the 21th century range between 2000 and 2021
- The years belonging to the 20th century range between 1970 and 1999

Now we are going to split the string into the different elements to transform them into a more intelligible format and merge them later.

In [9]:
months = {'Jan':'01', 'Feb':'02', 'Mar':'03', 'Apr':'04', 'May':'05', 'Jun':'06',
          'Jul':'07', 'Aug':'08', 'Sep':'09', 'Oct':'10', 'Nov':'11', 'Dec':'12'}

In [10]:
def date_reformat(date):
    
    """
    This function converts string dates in the dddd mm yy (e.g. '02nd Jul 98') format into string dates in the dd-mm-YYYY format ('02-07-1998').
    """
    
    output = np.nan
    
    if type(date) != float:
        
        day = date[:2]                                  # Days
        month = date[5:8]
        year = date[-2:]
        current_year = dt.datetime.today().year - 2000
        
        month = months[month]                           # Months

        if int(year) < current_year:                    # Years

            year = '20' + year                          # 21th century

        else:

            year = '19' + year                          # 20th century
            
        output = '-'.join([day, month, year])           # Merge into a single string in the dd-mm-YYYY format

    return output

In [11]:
vgc_data['Release'] = vgc_data['Release'].apply(date_reformat)  

In [12]:
vgc_data.sample(10)

Title Platform                   Publisher  \
3672                            NHL 2000       PS                   EA Sports   
53273           You Have to Win the Game       PC             Minor Key Games   
23786  American Conquest: Divided Nation       PC  CDV Software Entertainment   
22230                 4 in 1 Action Pack      PS3             SouthPeak Games   
17462                          Caesar IV       PC        Sierra Entertainment   
2401        LEGO Harry Potter: Years 1-4      PS3    Warner Bros. Interactive   
12473                           Venetica      PS3                       Atari   
35450             International Karate +       PS      Ignition Entertainment   
17608               The Raven Remastered      PS4                  THQ Nordic   
55795                               GRIS      PS4            Devolver Digital   

                       Developer  Total_Shipped  Total_Sales     Release  
3672                   EA Canada            NaN         0.87  30-09-1999  
53273                    Unknown            NaN          NaN  06-05-2012  
23786                    Unknown            NaN          NaN  27-02-2006  
22230                    Unknown            NaN         0.00  08-05-2012  
17462  Tilted Mill Entertainment            NaN         0.04  26-09-2006  
2401           Traveller's Tales            NaN         1.37  29-06-2010  
12473         DECK13 Interactive            NaN         0.12  11-01-2011  
35450     Ignition Entertainment            NaN          NaN  01-01-2003  
17608             KING Art Games            NaN         0.03  13-03-2018  
55795              Nomada Studio            NaN          NaN  26-11-2019

Now that there's nothing left to clean (in the formatting sense), let's check for NaN values to see what to do with them (as we have already seen many).

In [13]:
vgc_data.isna().sum()

Title                0
Platform             0
Publisher            0
Developer           17
Total_Shipped    55967
Total_Sales      40350
Release           4156
dtype: int64

There are so many NaNs in the dataframe we have to decide what to do with them!

- We are interested in the sales, so taking non-NaN `Total_Sales` values seems logical (other Sales columns have more NaNs, so it won't affect as much).
- Also, there are titles having NaN `Total_Sales` values but non-NaN `Total_Shipped` values.

Thus, we want to combine these two columns, and filling the NaN values.

However, to do so, first we have to check if there is any case in which both `Total_Sales` and `Total_Shipped` are non-NaN values.

We are going to fill the NaN values with zeros (we will consider 0 as not having a value in this case) and save them to a variable.

Then, we are going to perform the comparation `!= 0` to see how many non-zero values we have in each column.

Also, we cast the result to int so that we can sum them to check what we wanted to discover:

- If **sum = 0** : Both columns have either 0 or NaN.
- If **sum = 1** : Only one column have non-NaN or non-zero values.
- If **sum = 1** : Both columns have non-NaN and non-zero values.

In [14]:
vgc_data[vgc_data['Total_Sales'] == 0.34]

Title Platform  \
7099                       Kurushi Final: Mental Blocks       PS   
7100                           Risk / Battleship / Clue      GBA   
7101                                        Mini Ninjas      PS3   
7102           Pirates of the Caribbean: At World's End       DS   
7103                       SD Gundam G Generation World      PSP   
...                                                 ...      ...   
7213           Pirates of the Caribbean: At World's End     X360   
7214                       Delta Force: Black Hawk Down       XB   
7216                                             NHL 09      PS3   
7217                           Shaun White Snowboarding      PSP   
7218  World Championship Poker 2: Featuring Howard L...      PS2   

                        Publisher                       Developer  \
7099  Sony Computer Entertainment                            SCEI   
7100                    DSI Games                       Gravity-I   
7101            Eidos Interactive                  IO Interactive   
7102   Disney Interactive Studios             Amaze Entertainment   
7103                 Namco Bandai                      Tom Create   
...                           ...                             ...   
7213   Disney Interactive Studios  Eurocom Entertainment Software   
7214                    NovaLogic                    Climax Group   
7216              Electronic Arts                       EA Canada   
7217                      Ubisoft                Ubisoft Montreal   
7218          Crave Entertainment                   Point of View   

      Total_Shipped  Total_Sales     Release  
7099            NaN         0.34  01-08-1999  
7100            NaN         0.34  21-08-2005  
7101            NaN         0.34  08-09-2009  
7102            NaN         0.34  22-05-2007  
7103            NaN         0.34  24-02-2011  
...             ...          ...         ...  
7213            NaN         0.34  22-05-2007  
7214            NaN         0.34  26-07-2005  
7216            NaN         0.34  09-09-2008  
7217            NaN         0.34  16-11-2008  
7218            NaN         0.34  08-11-2005  

[100 rows x 7 columns]

In [15]:
total_sales = vgc_data['Total_Sales'].fillna(0)
total_shipped = vgc_data['Total_Shipped'].fillna(0)

In [16]:
zero_test = (total_sales != 0).astype(int) + (total_shipped != 0).astype(int)
zero_test.value_counts()

0    38279
1    21119
dtype: int64

This means that only one of the columns will hold a non-NaN or a non-zero value for the same data point.

In [17]:
zeros, ones = zero_test.value_counts()
zeros / (zeros + ones)

0.6444493080575103

And this means that 64.45% of the dataset would have zeros or NaNs in these two columns combined. 

Thus, in order to combine both columns and filling the NaNs, we will have to do the following:

- Save both columns to separate variables filling the NaNs with 0 (we can use the variables we used above).

- Perform a column-to-column addition.

- Remove the zeros.

- Get the median value.

- Sum `Total_sales` and `Total_shipped`, fill NaNs with the median value we calculated and save to a new column: `Sales`.

In [18]:
combined = total_sales + total_shipped
idx_zeros = list(combined[combined <= 0].index)
combined = combined.drop(index = idx_zeros)
med = combined.median()

# Median value we will use to fill the NaNs.
med

0.17

In [19]:
vgc_data['Sales'] = vgc_data[['Total_Shipped','Total_Sales']].sum(axis = 1, min_count = 1)\
                                                             .fillna(med)
vgc_data = vgc_data.drop(['Total_Shipped','Total_Sales'], axis = 1)

In [20]:
vgc_data.sample(10)

Title Platform  \
596                                   Final Fantasy X-2      PS2   
40750                  No Fear Downhill Mountain Biking       GB   
33665       Grand Theft Auto IV: The Ballad of Gay Tony      XBL   
23554                                             Alien     ACPC   
39300  Mobile Suit Gundam Seed: Tomo to Kimi to Koko de      GBA   
53911                    ACA NEOGEO SAMURAI SHODOWN III      PS4   
59069                                          Virginia      OSX   
35969                              Joe Montana Football       PC   
53529                                 Zoids Saga Fuzors      GBA   
28095              Command & Conquer: Covert Operations       PC   

                 Publisher         Developer     Release  Sales  
596            Square Enix       Square Enix  18-11-2003   5.50  
40750                  THQ       Codemasters  02-11-2001   0.17  
33665       Rockstar Games    Rockstar North  29-10-2009   0.17  
23554             Sega AM7  Concept Software  01-01-1985   0.17  
39300               Bandai            Bandai  13-05-2004   0.17  
53911  Hamster Corporation   SNK Corporation  19-04-2018   0.17  
59069              Unknown    Variable State         NaN   0.17  
35969                 Sega          MindSpan  01-01-1990   0.17  
53529     Tomy Corporation            Amedio  16-12-2004   0.17  
28095   Virgin Interactive  Westwood Studios  30-04-1996   0.17

---

We will strip `vgc_data` from the IPs (`Platform` = `Series`) and save them into a separate dataframe in case we happen to use it in the future.

In [21]:
vgc_ips = pd.DataFrame()

In [22]:
ips = vgc_data[vgc_data['Platform'] == 'Series']

In [23]:
ips = ips.drop(['Platform'], axis = 1)
ips

Title       Publisher  \
0                Tetris         Unknown   
1           Super Mario        Nintendo   
2               Pokemon        Nintendo   
3      Grand Theft Auto  Rockstar Games   
4                  FIFA       EA Sports   
...                 ...             ...   
56413           Lumines         Unknown   
57160      Perfect Dark         Unknown   
57454         Red Steel         Unknown   
57939        Sly Cooper         Unknown   
58252    Super Meat Boy       Team Meat   

                                   Developer     Release   Sales  
0                            Alexey Pajitnov  06-06-1984  496.00  
1                                   Nintendo  20-07-1983  372.86  
2                                 Game Freak  28-09-1998  369.88  
3                             Rockstar North  27-03-1998  335.00  
4      Extended Play Productions (1991-1997)  15-12-1993  325.00  
...                                      ...         ...     ...  
56413                        Q Entertainment         NaN    0.17  
57160                                   Rare         NaN    0.17  
57454                          Ubisoft Paris         NaN    0.17  
57939                           Sucker Punch         NaN    0.17  
58252                              Team Meat  20-10-2010    0.17  

[346 rows x 5 columns]

In [24]:
ips.to_csv("01_vgc_ips.csv", encoding='utf-8', index=False)

In [25]:
ip_idx = list(vgc_data[vgc_data['Platform'] == 'Series'].index)

In [26]:
vgc_data = vgc_data.drop(index = ip_idx)

We will save the result into a csv.

In [27]:
vgc_data.to_csv("01_vgc_clean.csv", encoding='utf-8', index=False)

## Some observations on vgc_data

### No. of Platforms

In [28]:
vgc_data['Platform'].value_counts()

PC      11653
PS2      3566
DS       3288
PS       2706
PS4      2255
        ...  
CD32        3
S32X        3
BBCM        1
Aco         1
C128        1
Name: Platform, Length: 78, dtype: int64

There are **77 platforms** (`Series` refers to the IP, and `All` refers to 'multiplatform')

---

### No. of Publishers

In [29]:
vgc_data['Publisher'].value_counts()

Unknown                5948
Sega                   2130
Ubisoft                1583
Electronic Arts        1532
Activision             1530
                       ... 
Culture Publishers        1
Simpy Entertainment       1
Dan Da Rocha              1
Rogue Play, Inc.          1
Logic Artists             1
Name: Publisher, Length: 3295, dtype: int64

There are **3294 publishers** (one is Unkown)

---

### No. of Developers

In [30]:
vgc_data['Developer'].value_counts()

Unknown                  4433
Konami                    932
Sega                      865
Capcom                    763
Namco                     427
                         ... 
Trans-Pegasus Limited       1
Mark Healey                 1
Case in Point Studios       1
JDPCreations                1
Punchline                   1
Name: Developer, Length: 8420, dtype: int64

There are **8441 developers** in total.

---

### **Decisions yet to make**
---

- **What do we do with the empty 38279 `Sales` values?**

    - Discard them.
    - Fill them with values -> which values?
    
    
- **What do we do with the 4156 NaN values from the `Release` column?**
    
    - If the release date is unknown it is higly likely that the title is very old (we should check them).
    
    - If it is a very old game, we shuld discard them as they will not hold much information.
    
    - If it is an IP, it will depend on what we decide to do with `All` platform.
    
    
- **What do we do with the 17 NaN velues from the `Developer` column?**

    - Probably same as `Release` column.


- **¿Qué hacemos con los juegos que tienen clasificación de 'All' y 'Series' en la columna `Platform`?**

    - The sum of sales in the different platforms does not usually match the `All` and `Series` total sales.
    
    - We have to take into account that the RAWG dataset has platforms grouped from the beginning and that there is no `All` or `Series` category.
        - Should we unpack those raws? -> it already has >500k data points so it does not seem to be the most intelligent thing to do.
        

- **How are we going to transform these categorical variables into featrues (numerical/boolan)?**

    - Platform: There are 77+1 platforms. Its dimensions is too high to do the usual One-Hot Encoding.
    
    - Publisher y Developer: There are 2 orders of magnitude above the number of platforms, thus we have to see what do we do with them to avoid the *Curse of Dimensionality*.

---

# Cleaning rawg_data

Let's see how many games are in the TBA state (`tba`) column

In [31]:
rawg_data[rawg_data['tba']==True]

id                                           slug  \
1581    303576            vampire-the-masquerade-bloodlines-2   
1999     25829                                       gauntlet   
2687    272483                          whispers-of-a-machine   
3034     58751                                  halo-infinite   
3295     11025              totally-accurate-battle-simulator   
...        ...                                            ...   
513054   43660                                           anne   
514307   41141                                      jack-claw   
514898   39758                       bonk-brink-of-extinction   
514999   39514  pirates-of-the-caribbean-armada-of-the-damned   
516213   35220                               good-cop-bad-cop   

                                                  name   released   tba  \
1581            Vampire: The Masquerade - Bloodlines 2 2022-12-31  True   
1999                                          Gauntlet 2009-02-28  True   
2687                             Whispers of a Machine 2019-04-16  True   
3034                                     Halo Infinite 2021-12-31  True   
3295                 Totally Accurate Battle Simulator        NaT  True   
...                                                ...        ...   ...   
513054                                        A.N.N.E.        NaT  True   
514307                                       Jack Claw        NaT  True   
514898                       Bonk: Brink of Extinction        NaT  True   
514999  Pirates of the Caribbean: Armada of the Damned        NaT  True   
516213                                Good Cop Bad Cop        NaT  True   

        metacritic  suggestions_count    updated  \
1581           NaN                563 2021-02-23   
1999           NaN                541 2019-09-17   
2687          77.0                431 2019-09-10   
3034           NaN                539 2020-11-29   
3295           NaN                208 2020-08-11   
...            ...                ...        ...   
513054         NaN                206 2019-01-09   
514307         NaN                385 2020-02-16   
514898         NaN                462 2020-06-19   
514999         NaN                665 2021-01-09   
516213         NaN                281 2019-03-13   

                                                platforms  \
1581    ['PC', 'PlayStation 4', 'PlayStation 5', 'Xbox...   
1999    ['Android', 'PlayStation 4', 'PC', 'Nintendo DS']   
2687                               ['iOS', 'macOS', 'PC']   
3034                ['PC', 'Xbox Series S/X', 'Xbox One']   
3295                                                   PC   
...                                                   ...   
513054        ['PC', 'macOS', 'Nintendo Switch', 'Linux']   
514307                                                 PC   
514898               ['Xbox 360', 'Wii', 'PlayStation 3']   
514999                ['Xbox 360', 'PlayStation 3', 'PC']   
516213                                      PlayStation 2   

                                       genres  \
1581                        ['Action', 'RPG']   
1999                        ['Action', 'RPG']   
2687         ['Adventure', 'Indie', 'Puzzle']   
3034       ['Action', 'Shooter', 'Adventure']   
3295      ['Strategy', 'Simulation', 'Indie']   
...                                       ...   
513054  ['Action', 'Adventure', 'Platformer']   
514307                                 Action   
514898                                 Arcade   
514999                      ['Action', 'RPG']   
516213                ['Action', 'Adventure']   

                                   stores  \
1581       ['Steam', 'Epic Games', 'GOG']   
1999    ['Google Play', 'Nintendo Store']   
2687        ['App Store', 'Steam', 'GOG']   
3034                                Steam   
3295              ['Epic Games', 'Steam']   
...                                   ...   
513054                                NaN   
514307                 

There are 2435 titles in the TBA state. However, we can see that there are games in that stare whose `release` is **in the past**.

Let's check how many future releases are in this dataset, using Feb 28th as the delimiter for 'future', as our VGChartz dataset has been retrieved that day.

In we get to improve the scraper in the future, we should save the date with something like the following:

```
today = dt.datetime.today().strftime("%Y-%m-%d")
```

In [32]:
rawg_data[rawg_data['released']>='2021-02-28']

id                                 slug  \
1581    303576  vampire-the-masquerade-bloodlines-2   
3010     45775                            biomutant   
3034     58751                        halo-infinite   
3255    287342                subnautica-below-zero   
3585    463723                            far-cry-6   
...        ...                                  ...   
516096   35712                             game-110   
516244   35133                             atriarch   
516325   34593                 french-street-racing   
516344   32973                             adellion   
516382   31929                            substream   

                                          name   released    tba  metacritic  \
1581    Vampire: The Masquerade - Bloodlines 2 2022-12-31   True         NaN   
3010                                 Biomutant 2021-05-25  False         NaN   
3034                             Halo Infinite 2021-12-31   True         NaN   
3255                    Subnautica: Below Zero 2021-05-14  False         NaN   
3585                                 Far Cry 6 2021-12-31  False         NaN   
...                                        ...        ...    ...         ...   
516096                                   Исход 2030-01-01  False         NaN   
516244                                Atriarch 2030-01-01  False         NaN   
516325                    French Street Racing 2030-01-01  False         NaN   
516344                                Adellion 2030-01-01  False         NaN   
516382                               Substream 2030-01-01  False         NaN   

        suggestions_count    updated  \
1581                  563 2021-02-23   
3010                  476 2021-01-26   
3034                  539 2020-11-29   
3255                  362 2021-02-25   
3585                  628 2021-02-15   
...                   ...        ...   
516096                378 2019-01-09   
516244                497 2019-01-09   
516325                377 2019-01-09   
516344                340 2019-01-09   
516382                 65 2019-01-09   

                                                platforms  \
1581    ['PC', 'PlayStation 4', 'PlayStation 5', 'Xbox...   
3010                  ['PlayStation 4', 'Xbox One', 'PC']   
3034                ['PC', 'Xbox Series S/X', 'Xbox One']   
3255                   ['macOS', 'PC', 'Nintendo Switch']   
3585    ['PlayStation 5', 'Xbox Series S/X', 'PlayStat...   
...                                                   ...   
516096                                                 PC   
516244                                                 PC   
516325                                                 PC   
516344                                                 PC   
516382                                                 PC   

                                              genres  \
1581                               ['Action', 'RPG']   
3010                               ['Action', 'RPG']   
3034              ['Action', 'Shooter', 'Adventure']   
3255                          ['Adventure', 'Indie']   
3585                           ['Action', 'Shooter']   
...                                              ...   
516096                                           RPG   
516244  ['RPG', 'Strategy', 'Massively Multiplayer']   
516325                          ['Racing', 'Arcade']   
516344              ['RPG', 'Massively Multiplayer']   
516382                                        Arcade   

                                                   stores  \
1581                       ['Steam', 'Epic Games', 'GOG']   
3010                                     ['GOG', 'Steam']   
3034                                                Steam   
3255            ['Nintendo Store', 'Epic Games', 'Steam']   
3585    ['Epic Games', 'Xbox Store', 'PlayStation Store']   
...                                                   ...   
516096                                                NaN   
516244         

We can see that there are far more games that should have the `tba` but they don't. Thus, we can drop the `tba` column as it gives non-accurate information.

Furthermore:

- We will also drop `id` as it's only useful if we are going to do requests to the API, but we are not allowed to do more than 20.000 requests a month, so nothing we can do with a dataset with >500k ids. 

- `updated` is a useless column, as our goal is not to evaluate the updates for a games' success (if we want to predict if a game is a success, it shouldn't even have an update).

- We will keep `slug` even though it seems useless, as the Metacritic page for games follows the `https://www.metacritic.com/game/PLATFORM/SLUG` format.

In [33]:
rawg_data = rawg_data.drop(['id','tba','updated'], axis = 1)
rawg_data.head()

slug                         name   released  \
0          grand-theft-auto-v           Grand Theft Auto V 2013-09-17   
1                    portal-2                     Portal 2 2011-04-18   
2     the-witcher-3-wild-hunt     The Witcher 3: Wild Hunt 2015-05-18   
3                 tomb-raider           Tomb Raider (2013) 2013-03-05   
4  the-elder-scrolls-v-skyrim  The Elder Scrolls V: Skyrim 2011-11-11   

   metacritic  suggestions_count  \
0        97.0                416   
1        95.0                582   
2        92.0                678   
3        86.0                664   
4        94.0                621   

                                           platforms  \
0  ['PC', 'Xbox Series S/X', 'PlayStation 5', 'Pl...   
1  ['Xbox One', 'PlayStation 3', 'PC', 'Xbox 360'...   
2  ['PC', 'Xbox One', 'Nintendo Switch', 'PlaySta...   
3  ['PC', 'PlayStation 4', 'PlayStation 3', 'Xbox...   
4  ['PC', 'PlayStation 3', 'Xbox 360', 'Nintendo ...   

                           genres  \
0         ['Action', 'Adventure']   
1           ['Shooter', 'Puzzle']   
2  ['Action', 'Adventure', 'RPG']   
3         ['Action', 'Adventure']   
4               ['Action', 'RPG']   

                                              stores  \
0  ['Epic Games', 'PlayStation Store', 'Xbox Stor...   
1  ['Xbox Store', 'Xbox 360 Store', 'PlayStation ...   
2  ['GOG', 'Xbox Store', 'Steam', 'PlayStation St...   
3  ['App Store', 'Google Play', 'PlayStation Stor...   
4  ['Xbox 360 Store', 'Nintendo Store', 'Steam', ...   

                                                tags   esrb_rating  
0  ['Singleplayer', 'Steam Achievements', 'Multip...        Mature  
1  ['Singleplayer', 'Steam Achievements', 'Multip...  Everyone 10+  
2  ['Singleplayer', 'Atmospheric', 'Full controll...        Mature  
3  ['Singleplayer', 'Multiplayer', 'Atmospheric',...        Mature  
4  ['Singleplayer', 'Steam Achievements', 'steam-...        Mature

In [34]:
rawg_data.shape[0]

521613

## Some observations on rawg_data

### No. of non-NaN values

In [35]:
rawg_data.notna().sum()

slug                 521611
name                 521611
released             497005
metacritic             4926
suggestions_count    521613
platforms            517536
genres               405457
stores               491543
tags                 459545
esrb_rating           57830
dtype: int64

We can see that there are very few `metacritic` values available, as well as kind of few values on `esrb_rating`.

We may opt to drop this column (`metacritic`) after retrieving the scores from the Metacritic web using a scraper.

---

### No. of NaN values

In [36]:
rawg_data.isna().sum()

slug                      2
name                      2
released              24608
metacritic           516687
suggestions_count         0
platforms              4077
genres               116156
stores                30070
tags                  62068
esrb_rating          463783
dtype: int64

---

We are dropping the 2 rows with `slug` and `name` missing, as we won't know which games they are.

In [37]:
rawg_data[rawg_data['slug'].isna()]

slug name   released  metacritic  suggestions_count  \
34946   NaN  NaN 2018-01-09         NaN                 72   
467110  NaN  NaN 2016-09-25         NaN                 67   

                       platforms                   genres   stores  \
34946   ['PC', 'macOS', 'Linux']  ['Adventure', 'Puzzle']  itch.io   
467110                        PC                Adventure  itch.io   

                                                     tags esrb_rating  
34946   ['2D', 'Pixel Graphics', 'Short', 'Multiple En...         NaN  
467110                ['Pixel Graphics', 'Point & Click']         NaN

In [38]:
rawg_data = rawg_data.drop(index = [34946, 467110])

In [39]:
rawg_data.shape[0]

521611

Saving the dataset into a csv.

In [40]:
rawg_data.to_csv("01_rawg_clean.csv", encoding='utf-8', index=False)

---
### **Decisions yet to make**
---

- How are we going to merge this dataset with vgc_data?

    - We are going to lowercase `name` and `title`, and if they match, they are added to the same row.
    
        - There are problems with the `platforms`, `genres`, `stores`, and `tags` columns.
        
        - Especially `platforms`, as vgc_data is separated into different platforms.
        

- How are we going to unpack the list-like columns? Are we going to use One-Hot Encoding?

    - We still **DON'T KNOW**


- What are we going to do with NaN values in some columns after the merge?

    - We still **DON'T KNOW**

In [41]:
a = vgc_data[vgc_data['Release'].isna()]

In [42]:
a

Title Platform Publisher  \
96         Call of Duty: Modern Warfare      All   Unknown   
117                    Human: Fall Flat      All   Unknown   
134                       Borderlands 2      All   Unknown   
135      Call of Duty: Modern Warfare 2      All   Unknown   
146                           Fallout 4      All   Unknown   
...                                 ...      ...       ...   
59375                     Yuppie Psycho      PS4   Unknown   
59377                            Zarvot       NS   Unknown   
59383                           Zombeer     WiiU   Unknown   
59387               Zombie Army Trilogy       NS   Unknown   
59388  Zombie Brigade: No Brain No Gain     WiiU   Unknown   

                    Developer Release  Sales  
96              Infinity Ward     NaN  30.00  
117           No Brakes Games     NaN  25.00  
134          Gearbox Software     NaN  22.00  
135             Infinity Ward     NaN  22.00  
146     Bethesda Game Studios     NaN  20.00  
...                       ...     ...    ...  
59375           Baroque Decay     NaN   0.17  
59377         Snowhydra Games     NaN   0.17  
59383           Padaone Games     NaN   0.17  
59387  Rebellion Developments     NaN   0.17  
59388          Cypron Studios     NaN   0.17  

[4072 rows x 6 columns]

In [43]:
a[a['Platform']=='All']

Title Platform Publisher  \
96       Call of Duty: Modern Warfare      All   Unknown   
117                  Human: Fall Flat      All   Unknown   
134                     Borderlands 2      All   Unknown   
135    Call of Duty: Modern Warfare 2      All   Unknown   
146                         Fallout 4      All   Unknown   
...                               ...      ...       ...   
58487  The Elder Scrolls IV: Oblivion      All   Unknown   
58810                   Torchlight II      All   Unknown   
59119               Warriors Orochi 2      All   Unknown   
59121               Warriors Orochi Z      All   Unknown   
59354           Yomawari: Night Alone      All   Unknown   

                   Developer Release  Sales  
96             Infinity Ward     NaN  30.00  
117          No Brakes Games     NaN  25.00  
134         Gearbox Software     NaN  22.00  
135            Infinity Ward     NaN  22.00  
146    Bethesda Game Studios     NaN  20.00  
...                      ...     ...    ...  
58487  Bethesda Game Studios     NaN   0.17  
58810            Runic Games     NaN   0.17  
59119            Omega Force     NaN   0.17  
59121            Omega Force     NaN   0.17  
59354   Nippon Ichi Software     NaN   0.17  

[623 rows x 6 columns]

In [44]:
a[a['Sales']>0]

Title Platform Publisher  \
96         Call of Duty: Modern Warfare      All   Unknown   
117                    Human: Fall Flat      All   Unknown   
134                       Borderlands 2      All   Unknown   
135      Call of Duty: Modern Warfare 2      All   Unknown   
146                           Fallout 4      All   Unknown   
...                                 ...      ...       ...   
59375                     Yuppie Psycho      PS4   Unknown   
59377                            Zarvot       NS   Unknown   
59383                           Zombeer     WiiU   Unknown   
59387               Zombie Army Trilogy       NS   Unknown   
59388  Zombie Brigade: No Brain No Gain     WiiU   Unknown   

                    Developer Release  Sales  
96              Infinity Ward     NaN  30.00  
117           No Brakes Games     NaN  25.00  
134          Gearbox Software     NaN  22.00  
135             Infinity Ward     NaN  22.00  
146     Bethesda Game Studios     NaN  20.00  
...                       ...     ...    ...  
59375           Baroque Decay     NaN   0.17  
59377         Snowhydra Games     NaN   0.17  
59383           Padaone Games     NaN   0.17  
59387  Rebellion Developments     NaN   0.17  
59388          Cypron Studios     NaN   0.17  

[4045 rows x 6 columns]